### ISSUE TITLE
git lfs migrate import performance issues with version 3.0.2

### ISSUE THREAD

**Issue by rolandas-rimkus** [git-lfs/git-lfs#4750]
**Describe the bug**
Running **git lfs migrate import** using version 3.0.2 is around 60 times slower compared to 3.0.1. Git repository with 5941 commits takes around 9 minutes to complete using version 3.0.1 and using 3.0.2 only managed to reach 60% of the commits in 6 hours of running.

**To Reproduce**
1. Download and install Git version 2.34.1 or 2.34.0 from https://github.com/git-for-windows/git/releases
2. Run **git lfs migrate import --yes --include="..."** on a Git repository
3. Download 3.0.1 version of Git LFS from https://github.com/git-lfs/git-lfs/releases/download/v3.0.1/git-lfs-windows-amd64-v3.0.1.zip
4. Apply the git-lfs.exe file on top of mingw64\bin\git-lfs.exe in the install location of Git
5. Repeat step 2 on a fresh Git repository

**Expected behavior**
Time taken for git lfs migrate import command to run should be identical with v3.0.1 and v3.0.2.

**System environment**
Windows 10 v1909

**Output of `git lfs env`**
git-lfs/3.0.2 (GitHub; windows amd64; go 1.17.2)
git version 2.34.0.windows.1

LocalWorkingDir=D:\Work\SubTest\Subsurface_Git-2.34.0
LocalGitDir=D:\Work\SubTest\Subsurface_Git-2.34.0\.git
LocalGitStorageDir=D:\Work\SubTest\Subsurface_Git-2.34.0\.git
LocalMediaDir=D:\Work\SubTest\Subsurface_Git-2.34.0\.git\lfs\objects
LocalReferenceDirs=
TempDir=D:\Work\SubTest\Subsurface_Git-2.34.0\.git\lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=D:\Work\SubTest\Subsurface_Git-2.34.0\.git\lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXE="C:\Git\Git-2.34.0\cmd\git.exe"
GIT_EXEC_PATH=C:/Git/Git-2.34.0/mingw64/libexec/git-core
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"

**Additional context**
This issue seems to be independent from Git version itself as I managed to reproduce the same issue with replacing the Git LFS delivered in Git versions 2.33.0, 2.33.1, 2.34.0 and 2.34.1 with versions 2.13.3, 3.0.1 or 3.0.2 of Git LFS and all Git versions are affected when using 3.0.2.

**Comment by chrisd8088** [comment]
Hey, thanks for the report and the numbers.  I wonder if this is in any way related to the changes in #4671?  Would you be able to build a Git LFS version which reverted that PR from v3.0.2 and try your timing test with that custom build?

**Comment by rolandas-rimkus** [comment]
#4671 did seem like the only changes between 3.0.1 and 3.0.2 that would cause this kind of issue when I was looking through the changes today. I can try reverting and testing locally over the weekend.

**Comment by rolandas-rimkus** [comment]
Managed to compile git-lfs with and without #4671 and I could reproduce the performance issue. Digging a little deeper into it I found that the minimal revert required to partially resolve the performance issues is to revert deletion of these lines under git/githistory/rewriter.go:
```
		if cached := r.uncacheEntry(entry); cached != nil {
			entries = append(entries, copyEntryMode(cached,
				entry.Filemode))
			continue
		}
```

So my best guess is that the new caching using blobCache under commands/command_migrate_import.go is not as efficient as the original one.

The performance I am getting right now is:
- 5:29 min with reverting everything in #4671
- 7:43 min with the minimal revert for git/githistory/rewriter.go
- 480 min without reverting

Here is the example output from the imports:
```
git version 2.34.0.windows.1
git-lfs/3.0.2 (GitHub; windows amd64; go 1.17.3)
Cloning into 'D:\Work\SubTest\Subsurface_Git-2.34.0-Lfs-Debug'...
remote: Enumerating objects: 77838, done.
remote: Counting objects: 100% (77838/77838), done.
remote: Compressing objects: 100% (70912/70912), done.
Receiving objects: 100% (77838/77838), 283.61 MiB | 7.67 MiB/s, done.
remote: Total 77838 (delta 57162), reused 0 (delta 0), pack-reused 0
Resolving deltas: 100% (57162/57162), done.
Updating files: 100% (3170/3170), done.
Updated git hooks.
Git LFS initialized.
migrate: Sorting commits: ..., done.
migrate: Rewriting commits: 100% (5941/5941), done.
  hg/Subsurface 8ccc1e3831a06cc0d39a5d148f3f1b6ecc724429 -> 9a3be8c7b907b1c6191f24e2cec012590e2ce17d
migrate: Updating refs: ..., done.
migrate: checkout: ..., done.
```

**Comment by chrisd8088** [comment]
That's great, thank you!  We can start taking a look next week at how much of an impact it would be to revert some or all of that PR.  It would obviously be ideal to retain the PR's functionality but without the performance hit; however, this sounds like a quite severe performance regression I don't think we can leave it unresolved.

**Comment by chrisd8088** [comment]
I just wanted to note here that while we haven't had a chance yet to look further, we will make sure to include a resolution to this performance problem in the next Git LFS release, which is slated for some time in January.

**Comment by andrein** [comment]
I am also experiencing this bug. Reverting git-lfs to 3.0.1 greatly speeds up the process.

**Comment by michael-house** [comment]
BTW, we ran into this and it was a significant blocker. Since this was our first time using the migration, it took a while to discover that migrating shouldn't take days (as it was for the size of our repo). It'd be great to address this issue ASAP and make it a bit more public as a known issue.

**Comment by Ikperfectik** [comment]
Wait sometimes

**Comment by bk2204** [comment]
We do plan on fixing this for 3.1, which we expect will be out in January, but we're all busy doing other things at the moment, so it will be fixed when we have sufficient time.  We are, of course, happy to accept a patch to fix this, provided it does not break the bugfix that was implemented (so a plain revert won't do).

**Comment by bk2204** [comment]
I've opened #4813 to fix this.  If folks are interested, we'd love to hear about how it works on your respective repositories.

**Comment by turboazot** [comment]
@rolandas-rimkus man, thanks for raising this issue. it helped me to understand why `git lfs migrate import` command was slow in ubuntu 22.04

Just in case if anybody will face this issue in ubuntu. I fixed it by upgrading git-lfs package from 3.0.2 to 3.3.0 like this:
```
curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash
sudo apt update
sudo apt install -y git-lfs
```

**Comment by tonit** [comment]
> @rolandas-rimkus man, thanks for raising this issue. it helped me to understand why `git lfs migrate import` command was slow in ubuntu 22.04
> 
> Just in case if anybody will face this issue in ubuntu. I fixed it by upgrading git-lfs package from 3.0.2 to 3.3.0 like this:
> 
> ```
> curl -s https://packagecloud.io/install/repositories/github/git-lfs/script.deb.sh | sudo bash
> sudo apt update
> sudo apt install -y git-lfs
> ```
> 
> Source: https://packagecloud.io/github/git-lfs/install

This actually just helped me a lot. The default git-lfs coming with Ubuntu 22.04 (latest LTS) is actually unusable for large rewrites. Thanks for pointing this out!
